### PR TITLE
Elasticsearch: Enable full range log volume histogram

### LINKED ISSUE #19949: Explore: Show complete timeline of logs with Loki
**What would you like to be added**:
Grafana's current integration with loki allows us to specify the number of logs to be returned. When this log limit has been reached, no more entries appear and the graph tails off. So at a glance it appears that logs are missing. Upping the log limit in the loki data source causes browser stability issues as all logs are dumped on screen.

A useful feature would be to return logs (within the window specified) but allow pagination to prevent browser issues. You could then allow limits to be set e.g. max per page.

It would also be helpful for the log limit to be configurable on each query. Having it in the data source makes it difficult to work with.

**Why is this needed**:
This would enhance the usability of loki and make it more viable as a replacement for other tools.

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

This is to enable log volume for Elastic ~and Loki~ with automatic loading after the query is run. 

*Update 2021/11/09: Loki requires `fullRangeLogsVolume` feature toggle to enable this feature.*

Changes:
* feature toggles are removed
* added extra resolution of 1ms when the selected time range is below 5 seconds is selected
* added `hint` param passed to Loki API when making a log volume query
* added documentation
* "Logs volume" renamed to "Log volume"

**Which issue(s) this PR fixes**:

Relates to #19949
Relates to #14481

**Special notes for your reviewer**: N/A

### CODE REVIEW

File: docs/sources/datasources/elasticsearch.md [hunk 1]
```diff
@@ -215,6 +215,8 @@ Once the result is returned, the log panel shows a list of log rows and a bar ch
 
 Note that the fields used for log message and level is based on an [optional data source configuration](#logs).
 
+Log volume panel above the log panel shows accumulated log distribution over time for all log queries. Results are split by level when level field is configured.
+
```
Comment (Elfo404) [inline docs/sources/datasources/elasticsearch.md:219]: The volume panel is shown only in Explore ATM, this is kind of confusing since it seems there's always a volume panel on top of the logs panel.

also, given this is an explore specific behaviour, maybe we'd be better off describing it in `docs/sources/explore/logs-integration.md` ?

wdyt?

File: docs/sources/datasources/loki.md [hunk 2]
```diff
@@ -66,11 +66,12 @@ You can use the Loki query editor to create log and metric queries.
 
 ### Log browser
 
-With Loki log browser you can easily navigate trough your list of labels and values and construct the query of your choice. Log browser has multi-step selection:
+With Loki log browser you can easily navigate through your list of labels and values and construct the query of your choice. Log browser has multi-step selection:
 
 1. Choose the labels you would like to consider for your search.
 2. Pick the values for selected labels. Log browser supports facetting and therefore it shows you only possible label combinations.
 3. Choose the type of query - logs query or rate metrics query. Additionally, you can also validate selector.
+4. Log volume panel above the log browser shows accumulated log distribution over time for all log queries.
```
Comment (Elfo404) [inline docs/sources/datasources/loki.md:74]: as far as I understood this section describes the steps required to create a query via the log browser, this line seems out of scope here

**Review by Elfo404 — COMMENTED** [review]
Seems to work fine 🚀 

i have some doubts regarding the docs changes though, so I'm not approving to wait for some input from @grafana/docs-squad

File: docs/sources/datasources/elasticsearch.md (hunk 1 above)
Comment (ifrost) [reply docs/sources/datasources/elasticsearch.md:219]: Good point! I'll move docs to https://grafana.com/docs/grafana/latest/explore/logs-integration/

File: docs/sources/datasources/loki.md (hunk 2 above)
Comment (ifrost) [reply docs/sources/datasources/loki.md:74]: I will ditch it and update https://grafana.com/docs/grafana/latest/explore/logs-integration/ instead 👍

File: docs/sources/datasources/elasticsearch.md (hunk 1 above)
Comment (ifrost) [reply docs/sources/datasources/elasticsearch.md]: Changed in 1492ff550e660aa6b199d83154efd22c7f1ba2bf

File: docs/sources/datasources/loki.md (hunk 2 above)
Comment (ifrost) [reply docs/sources/datasources/loki.md]: Changed in 1492ff550e660aa6b199d83154efd22c7f1ba2bf

File: public/app/plugins/datasource/loki/types.ts [hunk 3]
```diff
@@ -33,6 +34,8 @@ export interface LokiQuery extends DataQuery {
   resolution?: number;
   range?: boolean;
   instant?: boolean;
+  // extra info passed to Loki API when log volume query is run
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/types.ts:37]: I think this comment should be next to `hint` param in the `LokiRangeQueryRequest`

**Review by aocenas — APPROVED** [review]
Small nitpick regarding the comment, otherwise looks good.

File: public/app/plugins/datasource/loki/types.ts (hunk 3 above)
Comment (ifrost) [reply public/app/plugins/datasource/loki/types.ts]: Updated in eb618d74b4d6caf72e789f02097d42fd00098520

File: public/app/features/explore/LogsVolumePanel.tsx [hunk 4]
```diff
@@ -60,14 +60,14 @@ export function LogsVolumePanel(props: Props) {
   if (zoomRatio !== undefined && zoomRatio < 1) {
     zoomLevelInfo = (
       <>
-        <span className={styles.zoomInfo}>Reload logs volume</span>
+        <span className={styles.zoomInfo}>Reload log volume</span>
         <Button size="xs" icon="sync" variant="secondary" onClick={onLoadLogsVolume} />
       </>
     );
```
Comment (Elfo404) [inline public/app/features/explore/LogsVolumePanel.tsx:66]: as reported by Accessibility insights, this button is lacking a label. we could use the text next it as a label:

```suggestion
    zoomLevelInfo = (
      <label className={styles.zoomInfo}>
        Reload log volume&nbsp;
        <Button size="xs" icon="sync" variant="secondary" onClick={onLoadLogsVolume} />
      </label>
    );
```

(in this case i'd also add `cursor: pointer` to `styles.zoomInfo`)

Comment (Elfo404) [reply public/app/features/explore/LogsVolumePanel.tsx:66]: actually, we should probably be using grafana-ui:
```
zoomLevelInfo = (
      <InlineField className={styles.zoomInfo} label="Reload log volume" transparent>
        <Button size="xs" icon="sync" variant="secondary" onClick={onLoadLogsVolume} id="reload-volume" />
      </InlineField>
    );
 ```

in this case we can also remove `styles.zoomInfo`

Comment (ifrost) [reply public/app/features/explore/LogsVolumePanel.tsx]: Good point. Updated in 38f912d246e847c6cf7bc395d0ef20bfca5d6b74

File: docs/sources/explore/logs-integration.md [hunk 5]
```diff
@@ -21,7 +21,11 @@ During an infrastructure monitoring and incident response, you can dig deeper in
 
 ### Logs visualization
 
-Results of log queries are shown as histograms in the graph and individual logs are displayed below. If the data source does not send histogram data for the requested time range, the logs model computes a time series based on the log row counts bucketed by an automatically calculated time interval and the start of the histogram is then anchored by the first log row's timestamp from the result. The end of the time series is anchored to the time picker's **To** range.
+Results of log queries are shown as histograms in the graph and individual logs are displayed below.
+
+If the data source supports full range log volume histogram, the graph with log distribution for all entered log queries is shown automatically. This feature is currently supported by Loki and Elasticsearch data sources.
+
+If the data source does not support loading full range log volume histogram, the logs model computes a time series based on the log row counts bucketed by an automatically calculated time interval and the start of the histogram is then anchored by the first log row's timestamp from the result. The end of the time series is anchored to the time picker's **To** range.
```
Comment (achatterjee-grafana) [inline docs/sources/explore/logs-integration.md:28]:
```suggestion
Results of log queries are shown as histograms in the graph and individual logs are explained in the following sections.

If the data source supports a full range log volume histogram, the graph with log distribution for all entered log queries is shown automatically. This feature is currently supported by Loki and Elasticsearch data sources.

If the data source does not support loading full range log volume histogram, the logs model computes a time series based on the log row counts bucketed by an automatically calculated time interval, and the first log row's timestamp then anchors the start of the histogram from the result. The end of the time series is anchored to the time picker's **To** range.
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc review.

**Review by ivanahuckova — COMMENTED** [review]
LGTM from me as well. 🙂

**Review by Elfo404 — APPROVED** [review]
LGTM! 🪨👽!

**Comment by ifrost** [comment]
Loki's log volume was put back behind the feature toggle in 411e49da536b800f358bc8f6bd725ffe9a43a5aa. Test and docs were updated.
